### ISSUE TITLE
Extension regarding getDirectionalDerivatives()

### ISSUE THREAD

**Issue by jph-tavella** [modelica/fmi-standard#645]
This issue is not for clarification only but for extension of the possibilities for the use of the primitive getDirectionalDerivative(), also assuming its new name will be getDirectionalDerivatives().

The use case is the co-initialization of strongly coupled FMUs as we can see in power flow (electrical domain).
To success an electrical power flow, we have to calculate Jacobian for each FMU.  But with the current specification of getDirectionalDerivatives(), it is not possible to get the Jacobian in one call of the function, only one call per line (getting the derivatives of all outputs regarding one input).
As the getDirectionalDerivatives() is not simply a getter, it is not possible to calculate all the lines of the matrix in parallel (e.g. from Dymola FMUs). And of course serializing the calls to the primitive is not efficient.

For this reason, we propose to improve the possibilities for calling the primitive getDirectionalDerivatives() in order to allow getting with one call all the derivatives of a vector of outputs with respect to a vector of inputs, not restricting the current usage.

With FMI 2.0 here are two examples of using the primitive to get all the derivatives of a vector of outputs (o1, o2, o3) with respect to all inputs (i1, i2). In both case two calls are necessary.
 
// Example 1: two sequential calls are required
getDirectionalDerivatives(sim, {ref(o1),ref(o2),ref(o3)}, 3, {ref(i1)}, 1, {1}, res1[])  // res1 = {do1/di1,do2/di1,do3/di1}  all the derivatives of outputs with respect to THE FIRST input i1
getDirectionalDerivatives(sim, {ref(o1),ref(o2),ref(o3)}, 3, {ref(i2)}, 1, {1}, res2[])  // res2 = {do1/di2,do2/di2,do3/di2}  all the derivatives of outputs with respect to THE SECOND input i2

// Example 2: two sequential calls are required
getDirectionalDerivatives(sim, {ref(o1),ref(o2),ref(o3)}, 3, {ref(i1),ref(i2)}, 2, {1,0}, res1[]) // res1 = {do1/di1,do2/di1,do3/di1} all the derivatives of outputs with respect to THE FIRST input i1
getDirectionalDerivatives(sim, {ref(o1),ref(o2),ref(o3)}, 3, {ref(i1),ref(i2)}, 2, {0,1}, res1[]) // res2 = {do1/di2,do2/di2,do3/di2} all the derivatives of outputs with respect to THE SECOND input i2
  
FMI 3.0 is a great opportunity to extend the possibilities for calling the getDirectionalDerivatives() and get a better batch calling for this primitive.
// Example: only one call is required 
getDirectionalDerivatives(sim, {ref(o1),ref(o2),ref(o3)}, 3, {ref(i1),ref(i2)}, 2, {{1,0},{0,1}}, res[[]])  // res = {{do1/di1,do2/di1,do3/di1},{do1/di2,do2/di2,do3/di2}} all the derivatives of outputs with respect to all inputs
The idea is to provide for the seed a matrix adapted as a vector indicating the desired derivatives. Then, the result would be another matrix adapted as a vector. At this point the idea is to serialize and deserialize matrices as vectors to enhance the communication of dynamic libraries with programming languages.

Of course it is still possible to use the primitive as previously. Example:
getDirectionalDerivatives(sim, {ref(o1),ref(o2),ref(o3)}, 3, {ref(i1)}, 1, {{1}}, res1[[]])          // res1 = {{do1/di1,do2/di1,do3/di1}} all the derivatives of outputs with respect to THE FIRST input i1
getDirectionalDerivatives(sim, {ref(o1),ref(o2),ref(o3)}, 3, {ref(i2)}, 1, {{1}}, res2[[]])          // res2 = {{do1/di2,do2/di2,do3/di2}} all the derivatives of outputs with respect to THE SECOND input i2

**Comment by masoud-najafi** [comment]
Please note that in FMI-3.0, Known and Unknowns variables may be Vector or even Matrix. Adding this will increase the complexity of this API, already complex.

**Comment by jph-tavella** [comment]
@masoud-najafi: I see two difficulties with this primitive getDirectionalDerivatives().

First: the text in §2.1.9 in https://github.com/modelica/fmi-standard/blob/master/docs/2_1_common_api.adoc refers to arguments vrKnown, vrUnknown, nKnown, nUnknown, dvKnown, dvUnknown, nDvKnown, nDvunKnown while primitive description in https://github.com/modelica/fmi-standard/blob/master/headers/fmi3FunctionTypes.h refers to knowns, unknowns, nKowns, nUnkowns, deltaKnowns, deltaUnknowns, ndeltaKnowns, ndeltaUnknowns.
No sure it is not confusing; is fmi3FunctionTypes.h to be aligned to 2_1_common_api.adoc or the opposite?

Second: The issue I reported is not to handle vrKnown and vrUnknown as matrixes (I agree it should be an useless complication). The idea is simply to be able to define dvKnown and dvUnknown as a matrixes to get all derivatives in one call (refer to my initial example).

**Comment by chrbertsch** [comment]
This ticket seems to be a duplicate of #360. Please see also the discussion there.

I do not see an important benefit of one call instead of multiple calls, if a proper caching strategy is implemented, as the runtimetime will be in the calculation of the derivatives and not in the function call overhead.

**Comment by jph-tavella** [comment]
I'm on the same line as Sam.Lishak in his last comment on ticket #360 edited Oct 17, 2018.
My proposal and his last proposal are the same that is to say based on the concept of  "seed matrix". That way fmi2GetDirectionalDerivative could function the same as before, but would be extended such that if you pass in the identity matrix as your seed, you're returned a dense matrix which is the full Jacobian of the variables you requested.
Seem not being a hard to implement and not radically different than the current spririt of the primitive.

**Comment by chrbertsch** [comment]
Regular FMI Design webmeeting

@jph-tavella Please provide the concrete wording and API in a PR together with an example.

**Comment by chrbertsch** [comment]
Regular design meeting: 

Closing in favour of #360 (duplicate)
